### PR TITLE
WritableStream: Add tests for re-entrant calls inside strategySize

### PR DESCRIPTION
**Pull request by ricea**
A pathological implementation of strategy.size() can make re-entrant calls to WritableStream* methods. Implementations need to have standard behaviour in this case, so add tests to verify it.

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Chrome (unstable channel) #
Testing web-platform-tests at revision e59734f63d91f0aef5b3c718cd53349b23c0296f
Using browser at version 58.0.3004.3 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>8 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.sharedworker.html">/streams/writable-streams/reentrant-strategy.sharedworker.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.html">/streams/writable-streams/bad-strategies.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.html">/streams/writable-streams/reentrant-strategy.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.sharedworker.html">/streams/writable-streams/bad-strategies.sharedworker.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.serviceworker.https.html">/streams/writable-streams/bad-strategies.serviceworker.https.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Service worker test setup`                               | PASS    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.dedicatedworker.html">/streams/writable-streams/reentrant-strategy.dedicatedworker.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.serviceworker.https.html">/streams/writable-streams/reentrant-strategy.serviceworker.https.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `Service worker test setup`                                                      | PASS    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.dedicatedworker.html">/streams/writable-streams/bad-strategies.dedicatedworker.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
</details>

**Comment by w3c-bots** [comment]
# Firefox (nightly channel) #
Testing web-platform-tests at revision e59734f63d91f0aef5b3c718cd53349b23c0296f
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>8 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.sharedworker.html">/streams/writable-streams/reentrant-strategy.sharedworker.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.html">/streams/writable-streams/bad-strategies.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.html">/streams/writable-streams/reentrant-strategy.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.sharedworker.html">/streams/writable-streams/bad-strategies.sharedworker.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.serviceworker.https.html">/streams/writable-streams/bad-strategies.serviceworker.https.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Service worker test setup`                               | PASS    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.dedicatedworker.html">/streams/writable-streams/reentrant-strategy.dedicatedworker.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4830/streams/writable-streams/reentrant-strategy.serviceworker.https.html">/streams/writable-streams/reentrant-strategy.serviceworker.https.html</a></summary>


|                                     Subtest                                      | Results |
|----------------------------------------------------------------------------------|---------|
|                                                                                  | OK      |
| `original reader should error when new reader is created within strategy.size()` | FAIL    |
| `controller.error() should work when called from within strategy.size()`         | FAIL    |
| `releaseLock() should abort the write() when called within strategy.size()`      | FAIL    |
| `close() should work when called from within strategy.size()`                    | FAIL    |
| `Service worker test setup`                                                      | PASS    |
| `writes should be written in the standard order`                                 | FAIL    |
| `writer.write() promises should resolve in the standard order`                   | FAIL    |
| `abort() should work when called from within strategy.size()`                    | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4830/streams/writable-streams/bad-strategies.dedicatedworker.html">/streams/writable-streams/bad-strategies.dedicatedworker.html</a></summary>


|                          Subtest                          | Results |
|-----------------------------------------------------------|---------|
|                                                           | OK      |
| `reject any non-function value for strategy.size`         | FAIL    |
| `Writable stream: invalid strategy.size return value`     | FAIL    |
| `Writable stream: invalid strategy.highWaterMark`         | FAIL    |
| `Writable stream: throwing strategy.highWaterMark getter` | FAIL    |
| `Writable stream: throwing strategy.size method`          | FAIL    |
| `Writable stream: throwing strategy.size getter`          | FAIL    |

</details>
</details>

**Comment by ricea** [comment]
@tyoshino PTAL

File: streams/writable-streams/reentrant-strategy.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  let writer1;
+  let ws;
+  let writePromise2;
+  let closePromise;
+  let closedPromise2;
+  const strategy = {
+    size(chunk) {
+      if (chunk > 0) {
+        writer1.releaseLock();
+        const writer2 = ws.getWriter();
```
Comment (tyoshino) [inline streams/writable-streams/reentrant-strategy.js]: nice coverage :+1:

File: streams/writable-streams/reentrant-strategy.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        writer.write(chunk - 1);
+      }
+      return chunk;
+    }
+  };
+
+  const ws = recordingWritableStream({}, strategy);
+  writer = ws.getWriter();
+  return writer.write(2)
+      .then(() => Promise.resolve()) // See next test for why this is needed.
```
Comment (tyoshino) [inline streams/writable-streams/reentrant-strategy.js]: could you please elaborate this comment a bit more? "next test" is 'writer.write() promises should resolve in the standard order'?

i guess this is to run micro tasks here to run the queued sink.write(), but not sure how it's explained in the next test.

Comment (ricea) [inline streams/writable-streams/reentrant-strategy.js]: Sorry, that was leftover from when the behaviour was more pathological. I have removed it.

**Comment by ricea** [comment]
@tyoahino I messed up my last push. PTAL.

**Review by tyoshino — APPROVED** [review]

**Comment by tyoshino** [comment]
lgtm

**Pull request merged**